[Android]: running multiple tasks simultaneously throws  file not found exception.

## Steps to Reproduce
1. build add-to-app project with jenkins pipeline.
2. start multiple tasks at same time.
3. error occured.


## Logs

<details>
<summary>Logs</summary>

```
16:37:48  
16:37:48  FAILURE: Build failed with an exception.
16:37:48  
16:37:48  * What went wrong:
16:37:48  Execution failed for task ':connectivity:generateReleaseRFile'.
16:37:48  > A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
16:37:48     > /data/jenkins/.pub-cache/hosted/pub.flutter-io.cn/connectivity-3.0.6/android/build/intermediates/local_only_symbol_list/release/R-def.txt
16:37:48  
16:37:48  * Try:
16:37:48  Run with --info or --debug option to get more log output. Run with --scan to get full insights.
16:37:48  
16:37:48  * Exception is:
16:37:48  org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':connectivity:generateReleaseRFile'.
16:37:48  	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:205)
16:37:48  	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
16:37:48  	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:203)
16:37:48  	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
16:37:48  	...... more
16:37:48  Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
16:37:48  	at org.gradle.workers.internal.DefaultWorkerExecutor$WorkItemExecution.waitForCompletion(DefaultWorkerExecutor.java:354)
16:37:48  	at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:142)
16:37:48  	at org.gradle.internal.work.DefaultAsyncWorkTracker.access$000(DefaultAsyncWorkTracker.java:34)
16:37:48  	at org.gradle.internal.work.DefaultAsyncWorkTracker$1.run(DefaultAsyncWorkTracker.java:106)
16:37:48  	at org.gradle.internal.Factories$1.create(Factories.java:26)
16:37:48  	...... more
16:37:48  Caused by: java.nio.file.NoSuchFileException: /data/jenkins/.pub-cache/hosted/pub.flutter-io.cn/connectivity-3.0.6/android/build/intermediates/local_only_symbol_list/release/R-def.txt
16:37:48  	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
16:37:48  	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
16:37:48  	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
16:37:48  	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
16:37:48  	at java.nio.file.Files.newByteChannel(Files.java:361)
16:37:48  	at java.nio.file.Files.newByteChannel(Files.java:407)
16:37:48  	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
16:37:48  	at java.nio.file.Files.newInputStream(Files.java:152)
16:37:48  	at java.nio.file.Files.newBufferedReader(Files.java:2784)
16:37:48  	at java.nio.file.Files.lines(Files.java:3744)
16:37:48  	...... more
16:37:48  
16:37:48  
16:37:48  * Get more help at https://help.gradle.org
16:37:48  
16:37:48  Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
16:37:48  Use '--warning-mode all' to show the individual deprecation warnings.
16:37:48  See https://docs.gradle.org/6.1.1/userguide/command_line_interface.html#sec:command_line_warnings
16:37:48  
16:37:48  BUILD FAILED in 3m 54s
16:37:48  183 actionable tasks: 183 executed
```
</details>

### flutter doctor -v

```
-bash-4.2$ ./bin/flutter doctor -v
[✓] Flutter (Channel stable, 2.2.3, on Linux, locale en_US.UTF-8)
    • Flutter version 2.2.3 at /data/jenkins/flutter_sdk_2_0
    • Framework revision f4abaa0735 (4 months ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /data/jenkins/.android_build_HOME/android-sdk-linux
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /data/jenkins/.android_build_HOME/android-sdk-linux
    • Java binary at: /usr/bin/java
    • Java version OpenJDK Runtime Environment 18.9 (build 11.0.12+7-LTS)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[!] Connected device
    ! No devices available

! Doctor found issues in 3 categories.
```

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -40,17 +40,13 @@ String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolute
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { sp ->
-            //only flutter submodules.
             if (androidPlugins != null) {
                 String sanitizedPluginName = sp.name.replace(":", "");
-                //matching submodule, which one's name on the `pluginsFile`.
                 if (androidPlugins.name.contains(sanitizedPluginName)){
                     File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator + "plugins_build_output" + File.separator + sanitizedPluginName);
                     if (!androidPluginBuildOutputDir.exists()){
-                        //make dir if not exist.
                         androidPluginBuildOutputDir.mkdirs()
                     }
-                    //let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.
                     sp.buildDir = androidPluginBuildOutputDir
                 }
             }

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -35,7 +35,6 @@ if (pluginsFile.exists()) {
     }
 }
 
-String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { sp ->

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -9,16 +9,15 @@ import groovy.json.JsonSlurper
 
 def moduleProjectRoot = project(':flutter').projectDir.parentFile.parentFile
 
-def androidPlugins = null;
+def object = null;
 String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
 // Note: if this logic is changed, also change the logic in app_plugin_loader.gradle.
 def pluginsFile = new File(moduleProjectRoot, '.flutter-plugins-dependencies')
 if (pluginsFile.exists()) {
-    def object = new JsonSlurper().parseText(pluginsFile.text)
+    object = new JsonSlurper().parseText(pluginsFile.text)
     assert object instanceof Map
     assert object.plugins instanceof Map
     assert object.plugins.android instanceof List
-    androidPlugins = object.plugins.android
     // Includes the Flutter plugins that support the Android platform.
     object.plugins.android.each { androidPlugin ->
         assert androidPlugin.name instanceof String
@@ -40,15 +39,14 @@ String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolute
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { sp ->
-            if (androidPlugins != null) {
+            if (object != null && object.plugins.android.name.contains(sanitizedPluginName)) {
                 String sanitizedPluginName = sp.name.replace(":", "");
-                if (androidPlugins.name.contains(sanitizedPluginName)){
-                    File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator + "plugins_build_output" + File.separator + sanitizedPluginName);
-                    if (!androidPluginBuildOutputDir.exists()){
-                        androidPluginBuildOutputDir.mkdirs()
-                    }
-                    sp.buildDir = androidPluginBuildOutputDir
+                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator
+                        + "plugins_build_output" + File.separator + sanitizedPluginName);
+                if (!androidPluginBuildOutputDir.exists()){
+                    androidPluginBuildOutputDir.mkdirs()
                 }
+                sp.buildDir = androidPluginBuildOutputDir
             }
         }
         def _mainModuleName = binding.variables['mainModuleName']

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -38,9 +38,9 @@ if (pluginsFile.exists()) {
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { sp ->
+            String sanitizedPluginName = sp.name.replace(":", "");
             if (object != null && object.plugins != null && object.plugins.android != null
                     && object.plugins.android.name.contains(sanitizedPluginName)) {
-                String sanitizedPluginName = sp.name.replace(":", "");
                 File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator
                         + "plugins_build_output" + File.separator + sanitizedPluginName);
                 if (!androidPluginBuildOutputDir.exists()) {

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -11,6 +11,7 @@ def moduleProjectRoot = project(':flutter').projectDir.parentFile.parentFile
 
 // Note: if this logic is changed, also change the logic in app_plugin_loader.gradle.
 def pluginsFile = new File(moduleProjectRoot, '.flutter-plugins-dependencies')
+List<String> pluginsNames = new ArrayList<>()
 if (pluginsFile.exists()) {
     def object = new JsonSlurper().parseText(pluginsFile.text)
     assert object instanceof Map
@@ -20,6 +21,7 @@ if (pluginsFile.exists()) {
     object.plugins.android.each { androidPlugin ->
         assert androidPlugin.name instanceof String
         assert androidPlugin.path instanceof String
+        pluginsNames.add(androidPlugin.name)
         // Skip plugins that have no native build (such as a Dart-only
         // implementation of a federated plugin).
         def needsBuild = androidPlugin.containsKey('native_build') ? androidPlugin['native_build'] : true
@@ -33,8 +35,20 @@ if (pluginsFile.exists()) {
     }
 }
 
+String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
+        //let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.
+        p.subprojects { sp ->
+            if (pluginsNames.contains(sp.name.replace(":", ""))){
+                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator + "plugins_build_output" + File.separator + sp.name.replace(":", ""));
+                if (!androidPluginBuildOutputDir.exists()){
+                    //make dir if not exist.
+                    androidPluginBuildOutputDir.mkdirs()
+                }
+                sp.buildDir = androidPluginBuildOutputDir
+            }
+        }
         def _mainModuleName = binding.variables['mainModuleName']
         if (_mainModuleName != null && !_mainModuleName.empty) {
             p.ext.mainModuleName = _mainModuleName

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -35,18 +35,12 @@ if (pluginsFile.exists()) {
     }
 }
 
-String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
-        //let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.
+        //let the plugin's output generated in "${rootProject.buildDir}/${androidPlugin.name}".
         p.subprojects { sp ->
             if (pluginsNames.contains(sp.name.replace(":", ""))){
-                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator + "plugins_build_output" + File.separator + sp.name.replace(":", ""));
-                if (!androidPluginBuildOutputDir.exists()){
-                    //make dir if not exist.
-                    androidPluginBuildOutputDir.mkdirs()
-                }
-                sp.buildDir = androidPluginBuildOutputDir
+                sp.buildDir = "${rootProject.buildDir}/${sp.name.replace(":", "")}"
             }
         }
         def _mainModuleName = binding.variables['mainModuleName']

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -21,6 +21,7 @@ if (pluginsFile.exists()) {
     object.plugins.android.each { androidPlugin ->
         assert androidPlugin.name instanceof String
         assert androidPlugin.path instanceof String
+        // save plugin's name for modifying it's buildDir location. see issue #92252 for detail.
         pluginsNames.add(androidPlugin.name)
         // Skip plugins that have no native build (such as a Dart-only
         // implementation of a federated plugin).
@@ -45,6 +46,8 @@ gradle.getGradle().projectsLoaded { g ->
                     //make dir if not exist.
                     androidPluginBuildOutputDir.mkdirs()
                 }
+                // only the project object here can modify the `buildDir`.
+                // {@see https://github.com/flutter/flutter/pull/94645/files/7c77d3653709ccd1e2b45268b209f00f36d8470d#r819224682}
                 sp.buildDir = androidPluginBuildOutputDir
             }
         }

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -42,9 +42,9 @@ gradle.getGradle().projectsLoaded { g ->
             if (object != null && object.plugins != null && object.plugins.android != null
                     && object.plugins.android.name.contains(sanitizedPluginName)) {
                 String sanitizedPluginName = sp.name.replace(":", "");
-                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator
+                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator
                         + "plugins_build_output" + File.separator + sanitizedPluginName);
-                if (!androidPluginBuildOutputDir.exists()){
+                if (!androidPluginBuildOutputDir.exists()) {
                     androidPluginBuildOutputDir.mkdirs()
                 }
                 sp.buildDir = androidPluginBuildOutputDir

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -40,7 +40,7 @@ String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolute
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { sp ->
-            //only flutter 'add-to-app' project.
+            //only flutter submodules.
             if (androidPlugins != null) {
                 String sanitizedPluginName = sp.name.replace(":", "");
                 //matching submodule, which one's name on the `pluginsFile`.

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -37,16 +37,16 @@ if (pluginsFile.exists()) {
 
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
-        p.subprojects { sp ->
-            String sanitizedPluginName = sp.name.replace(":", "");
+        p.subprojects { subproject ->
+            String pluginName = subproject.name.replace(":", "");
             if (object != null && object.plugins != null && object.plugins.android != null
-                    && object.plugins.android.name.contains(sanitizedPluginName)) {
+                    && object.plugins.android.name.contains(pluginName)) {
                 File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator
-                        + "plugins_build_output" + File.separator + sanitizedPluginName);
+                        + "plugins_build_output" + File.separator + pluginName);
                 if (!androidPluginBuildOutputDir.exists()) {
                     androidPluginBuildOutputDir.mkdirs()
                 }
-                sp.buildDir = androidPluginBuildOutputDir
+                subproject.buildDir = androidPluginBuildOutputDir
             }
         }
         def _mainModuleName = binding.variables['mainModuleName']

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -9,20 +9,20 @@ import groovy.json.JsonSlurper
 
 def moduleProjectRoot = project(':flutter').projectDir.parentFile.parentFile
 
+def androidPlugins = null;
+String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
 // Note: if this logic is changed, also change the logic in app_plugin_loader.gradle.
 def pluginsFile = new File(moduleProjectRoot, '.flutter-plugins-dependencies')
-List<String> pluginsNames = new ArrayList<>()
 if (pluginsFile.exists()) {
     def object = new JsonSlurper().parseText(pluginsFile.text)
     assert object instanceof Map
     assert object.plugins instanceof Map
     assert object.plugins.android instanceof List
+    androidPlugins = object.plugins.android
     // Includes the Flutter plugins that support the Android platform.
     object.plugins.android.each { androidPlugin ->
         assert androidPlugin.name instanceof String
         assert androidPlugin.path instanceof String
-        // save plugin's name for modifying it's buildDir location. see issue #92252 for detail.
-        pluginsNames.add(androidPlugin.name)
         // Skip plugins that have no native build (such as a Dart-only
         // implementation of a federated plugin).
         def needsBuild = androidPlugin.containsKey('native_build') ? androidPlugin['native_build'] : true
@@ -36,19 +36,23 @@ if (pluginsFile.exists()) {
     }
 }
 
+String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
-        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
         p.subprojects { sp ->
-            if (pluginsNames.contains(sp.name.replace(":", ""))){
-                File androidPluginBuildOutputDir = new File(moduleProjectRoot.getAbsolutePath() + File.separator + "build" + File.separator + "host" + File.separator + sp.name.replace(":", ""));
-                if (!androidPluginBuildOutputDir.exists()){
-                    //make dir if not exist.
-                    androidPluginBuildOutputDir.mkdirs()
+            //only flutter 'add-to-app' project.
+            if (androidPlugins != null) {
+                String sanitizedPluginName = sp.name.replace(":", "");
+                //matching submodule, which one's name on the `pluginsFile`.
+                if (androidPlugins.name.contains(sanitizedPluginName)){
+                    File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator + "plugins_build_output" + File.separator + sanitizedPluginName);
+                    if (!androidPluginBuildOutputDir.exists()){
+                        //make dir if not exist.
+                        androidPluginBuildOutputDir.mkdirs()
+                    }
+                    //let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.
+                    sp.buildDir = androidPluginBuildOutputDir
                 }
-                // only the project object here can modify the `buildDir`.
-                // {@see https://github.com/flutter/flutter/pull/94645/files/7c77d3653709ccd1e2b45268b209f00f36d8470d#r819224682}
-                sp.buildDir = androidPluginBuildOutputDir
             }
         }
         def _mainModuleName = binding.variables['mainModuleName']

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -35,13 +35,12 @@ if (pluginsFile.exists()) {
     }
 }
 
-String flutterModulePath = project(':flutter').projectDir.parentFile.parentFile.getAbsolutePath()
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
         p.subprojects { sp ->
             if (pluginsNames.contains(sp.name.replace(":", ""))){
-                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + "build" + File.separator + "host" + File.separator + sp.name.replace(":", ""));
+                File androidPluginBuildOutputDir = new File(moduleProjectRoot.getAbsolutePath() + File.separator + "build" + File.separator + "host" + File.separator + sp.name.replace(":", ""));
                 if (!androidPluginBuildOutputDir.exists()){
                     //make dir if not exist.
                     androidPluginBuildOutputDir.mkdirs()

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -39,7 +39,8 @@ String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolute
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { sp ->
-            if (object != null && object.plugins.android.name.contains(sanitizedPluginName)) {
+            if (object != null && object.plugins != null && object.plugins.android != null
+                    && object.plugins.android.name.contains(sanitizedPluginName)) {
                 String sanitizedPluginName = sp.name.replace(":", "");
                 File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator
                         + "plugins_build_output" + File.separator + sanitizedPluginName);

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -38,11 +38,10 @@ if (pluginsFile.exists()) {
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
         p.subprojects { subproject ->
-            String pluginName = subproject.name.replace(":", "");
             if (object != null && object.plugins != null && object.plugins.android != null
-                    && object.plugins.android.name.contains(pluginName)) {
+                    && object.plugins.android.name.contains(subproject.name)) {
                 File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator
-                        + "plugins_build_output" + File.separator + pluginName);
+                        + "plugins_build_output" + File.separator + subproject.name);
                 if (!androidPluginBuildOutputDir.exists()) {
                     androidPluginBuildOutputDir.mkdirs()
                 }

--- a/packages/flutter_tools/gradle/module_plugin_loader.gradle
+++ b/packages/flutter_tools/gradle/module_plugin_loader.gradle
@@ -35,12 +35,18 @@ if (pluginsFile.exists()) {
     }
 }
 
+String flutterModulePath = project(':flutter').projectDir.parentFile.parentFile.getAbsolutePath()
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
-        //let the plugin's output generated in "${rootProject.buildDir}/${androidPlugin.name}".
+        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
         p.subprojects { sp ->
             if (pluginsNames.contains(sp.name.replace(":", ""))){
-                sp.buildDir = "${rootProject.buildDir}/${sp.name.replace(":", "")}"
+                File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + "build" + File.separator + "host" + File.separator + sp.name.replace(":", ""));
+                if (!androidPluginBuildOutputDir.exists()){
+                    //make dir if not exist.
+                    androidPluginBuildOutputDir.mkdirs()
+                }
+                sp.buildDir = androidPluginBuildOutputDir
             }
         }
         def _mainModuleName = binding.variables['mainModuleName']

--- a/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
+++ b/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
@@ -42,7 +42,7 @@ void main() {
 
     final String projectPath = fileSystem.path.join(tempDir.path, 'flutter_project');
 
-    File modulePubspec = fileSystem.file(fileSystem.path.join(projectPath, 'pubspec.yaml'));
+    final File modulePubspec = fileSystem.file(fileSystem.path.join(projectPath, 'pubspec.yaml'));
     String pubspecContent = modulePubspec.readAsStringSync();
     pubspecContent = pubspecContent.replaceFirst(
       'dependencies:',

--- a/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
+++ b/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
@@ -1,0 +1,69 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file_testing/file_testing.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/io.dart';
+import 'package:flutter_tools/src/cache.dart';
+
+import '../src/common.dart';
+import 'test_utils.dart';
+
+void main() {
+
+  late Directory tempDir;
+  late String projectPath;
+
+  setUp(()  {
+    Cache.flutterRoot = getFlutterRoot();
+    tempDir = createResolvedTempDirectorySync('android_plugin_new_output_dir_test.');
+  });
+
+  tearDown(() async {
+    tryToDelete(tempDir);
+  });
+
+  test("error logged when plugin's output dir was flutter_project/", () async {
+    final String flutterBin = fileSystem.path.join(
+      getFlutterRoot(),
+      'bin',
+      'flutter',
+    );
+    // create flutter module project
+    ProcessResult result = processManager.runSync(<String>[
+      flutterBin,
+      ...getLocalEngineArguments(),
+      'create',
+      '--template=module',
+      'flutter_project'
+    ], workingDirectory: tempDir.path);
+
+    projectPath = fileSystem.path.join(tempDir.path, 'flutter_project');
+
+    File modulePubspec = fileSystem.file(fileSystem.path.join(projectPath, 'pubspec.yaml'));
+    String pubspecContent = modulePubspec.readAsStringSync();
+    pubspecContent = pubspecContent.replaceFirst(
+      'dependencies:',
+      '''dependencies:
+  image_picker: any''',
+    );
+    modulePubspec.writeAsStringSync(pubspecContent);
+
+    // Run flutter build apk to build module example project
+    result = processManager.runSync(<String>[
+      flutterBin,
+      ...getLocalEngineArguments(),
+      'build',
+      'aar',
+      '--target-platform=android-arm',
+    ], workingDirectory: projectPath);
+
+    // Check outputDir existed
+    final Directory outputDir = fileSystem.directory(fileSystem.path.join(
+        projectPath, '.android', 'plugins_build_output', 'image_picker_android'
+    ));
+    expect(outputDir, exists);
+
+  });
+}

--- a/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
+++ b/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
@@ -25,7 +25,7 @@ void main() {
     tryToDelete(tempDir);
   });
 
-  test("error logged when plugin's output dir was flutter_project/", () async {
+  test("error logged when plugin's build output dir was not private.", () async {
     final String flutterBin = fileSystem.path.join(
       getFlutterRoot(),
       'bin',

--- a/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
+++ b/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
@@ -2,6 +2,8 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:math';
+
 import 'package:file_testing/file_testing.dart';
 import 'package:flutter_tools/src/base/file_system.dart';
 import 'package:flutter_tools/src/base/io.dart';
@@ -13,7 +15,6 @@ import 'test_utils.dart';
 void main() {
 
   late Directory tempDir;
-  late String projectPath;
 
   setUp(()  {
     Cache.flutterRoot = getFlutterRoot();
@@ -39,13 +40,14 @@ void main() {
       'flutter_project'
     ], workingDirectory: tempDir.path);
 
-    projectPath = fileSystem.path.join(tempDir.path, 'flutter_project');
+    final String projectPath = fileSystem.path.join(tempDir.path, 'flutter_project');
 
     File modulePubspec = fileSystem.file(fileSystem.path.join(projectPath, 'pubspec.yaml'));
     String pubspecContent = modulePubspec.readAsStringSync();
     pubspecContent = pubspecContent.replaceFirst(
       'dependencies:',
-      '''dependencies:
+      '''
+dependencies:
   image_picker: any''',
     );
     modulePubspec.writeAsStringSync(pubspecContent);
@@ -58,6 +60,8 @@ void main() {
       'aar',
       '--target-platform=android-arm',
     ], workingDirectory: projectPath);
+
+    log(result.exitCode);
 
     // Check outputDir existed
     final Directory outputDir = fileSystem.directory(fileSystem.path.join(

--- a/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
+++ b/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart
@@ -48,7 +48,7 @@ void main() {
       'dependencies:',
       '''
 dependencies:
-  image_picker: any''',
+  image_picker: 0.8.5+3''',
     );
     modulePubspec.writeAsStringSync(pubspecContent);
 

